core/src/main/java/com/linecorp/armeria/server/RoutingPredicate.java

```diff
@@ -108,37 +115,120 @@
     @VisibleForTesting
     static <T, U> RoutingPredicate<T> of(String predicateExpr,
                                          Function<String, U> nameConverter,
-                                         Function<U, Predicate<T>> containsPredicateFactory,
-                                         BiFunction<U, String, Predicate<T>> equalsPredicateFactory) {
+                                         Function<U, Predicate<T>> containsPredicate,
+                                         BiFunction<U, String, Predicate<T>> equalsPredicate) {
+        final RoutingPredicate<T> routingPredicate;
+        final RoutingPredicate<T> orRoutingPredicate = buildOrRoutingPredicate(predicateExpr,
+                                                                               nameConverter,
+                                                                               containsPredicate,
+                                                                               equalsPredicate);
+        if (orRoutingPredicate != null) {
+            routingPredicate = orRoutingPredicate;
+        } else {
+            routingPredicate = buildSingleExprRoutingPredicate(predicateExpr,
+                                                               nameConverter,
+                                                               containsPredicate,
+                                                               equalsPredicate);
+        }
+        checkArgument(routingPredicate != null,
+                      "Invalid predicate: %s (expected: '%s', '%s' or '%s')",
+                      predicateExpr, CONTAIN_PATTERN.pattern(), COMPARE_PATTERN.pattern(), "A combination of " +
+                              "the previous predicates separated by '||'");
+        return new RoutingPredicate<>(routingPredicate.name, routingPredicate.delegate);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    private static <T, U> RoutingPredicate<T> buildOrRoutingPredicate(
+            String predicateExpr, Function<String, U> nameConverter,
+            Function<U, Predicate<T>> containsPredicate,
+            BiFunction<U, String, Predicate<T>> equalsPredicate) {
+        if (predicateExpr.contains("||")) {
+            if (predicateExpr.endsWith("|")) {
+                throw new IllegalArgumentException("Invalid predicate: " + predicateExpr +
+                                                           "(expected: '" +
+                                                           TRAILING_PIPE_PATTERN.matcher(predicateExpr)
+                                                                                .replaceAll("") +
+                                                           "')");
+            }
```

Comment:
How about splitting first with `OR_SPLITTER` and raising an exception when it yields an empty string, so we guard against the cases such as `foo || || bar` (empty component in the middle) and `foo || ` (trailing space after `||`)?

Reply:
Done